inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/non_max_suppression/non_max_suppression_kernel_ref.cpp

```diff
@@ -0,0 +1,298 @@
+        jit.AddConstant(MakeJitConstant("IOU_THRESHOLD_VAL", "convert_float(iou_threshold[0])"));
+    } else {
+        jit.AddConstant(MakeJitConstant("IOU_THRESHOLD_VAL", params.iou_threshold));
+    }
+
+    if (params.score_threshold_type == NmsArgType::Input) {
+        jit.AddConstant(MakeJitConstant("SCORE_THRESHOLD_TYPE", GetInputTypeStr(params.GetIndexScoreThreshold())));
+        jit.AddConstant(MakeJitConstant("SCORE_THRESHOLD_VAL", "convert_float(score_threshold[0])"));
+    } else {
+        jit.AddConstant(MakeJitConstant("SCORE_THRESHOLD_VAL", params.score_threshold));
+    }
+
+    if (params.soft_nms_sigma_type == NmsArgType::Input) {
+        jit.AddConstant(MakeJitConstant("SOFT_NMS_SIGMA_TYPE", GetInputTypeStr(params.GetIndexSoftNmsSigma())));
+        jit.AddConstant(MakeJitConstant("SOFT_NMS_SIGMA_VAL", "convert_float(soft_nms_sigma[0])"));
+    } else {
+        jit.AddConstant(MakeJitConstant("SOFT_NMS_SIGMA_VAL", params.soft_nms_sigma));
+    }
+
+    if (params.has_second_output) {
+        jit.AddConstant(MakeJitConstant("SECOND_OUTPUT_TYPE", GetInputTypeStr(params.GetIndexSecondOutput())));
+        jit.AddConstant(MakeJitConstant("TO_SECOND_OUTPUT_TYPE", GetToInputTypeStr(params.GetIndexSecondOutput())));
+    }
+
+    if (params.has_third_output) {
+        jit.AddConstant(MakeJitConstant("THIRD_OUTPUT_TYPE", GetInputTypeStr(params.GetIndexThirdOutput())));
+        jit.AddConstant(MakeJitConstant("TO_THIRD_OUTPUT_TYPE", GetToInputTypeStr(params.GetIndexThirdOutput())));
+    }
+
+    return jit;
+}
+
+static inline int GetPartitionStep(int localWorkItemNum) {
+    int step_size = 0;
+    for (int temp = localWorkItemNum; temp > 1; temp /= 2) {
+        step_size++;
+    }
+    return step_size;
+}
+
+static inline size_t GetOptimalLocalClassSize(std::vector<size_t> gws, const EngineInfo& info) {
+    const size_t optimal_values[] = {16, 8, 7, 6, 5, 4, 2, 1};
+    const size_t splitNum = gws[2];
+    const size_t globalClassNum = gws[1];
+    const auto rest_lws = info.maxWorkGroupSize / splitNum;
+    size_t lws_idx = 0;
+    while (rest_lws < optimal_values[lws_idx])
+        lws_idx++;
+    while (globalClassNum % optimal_values[lws_idx])
+        lws_idx++;
+
+    return optimal_values[lws_idx];
+}
+
+NonMaxSuppressionKernelRef::DispatchData SetDefault(const non_max_suppression_params& params, int idx) {
+    NonMaxSuppressionKernelRef::DispatchData dispatchData;
+
+    const auto& input = params.inputs[1];
+    if (idx == 0) {
+        dispatchData.gws = {input.Batch().v, input.Feature().v, params.engineInfo.maxWorkGroupSize};
+        dispatchData.lws = {1, 1, params.engineInfo.maxWorkGroupSize};
+    } else if (idx == 1) {
+        const size_t kSplitNum = 16;
+        dispatchData.gws = {input.Batch().v, input.Feature().v, kSplitNum};
+        const size_t kClassSize = GetOptimalLocalClassSize(dispatchData.gws, params.engineInfo);
+        dispatchData.lws = {1, kClassSize, kSplitNum};
```

Comment:
TGL/DG1 support bigger work groups (512), but here you can have up to 256 only. Won't bigger WG size improve the performance on gen12lp?